src/one/nio/ws/WebSocketSession.java

```diff
@@ -0,0 +1,211 @@
+        if (currentTime - lastAccessTime < keepAlive) {
+            return ACTIVE;
+        }
+
+        try {
+            if (wasSelected) {
+                sendMessage(PingMessage.EMPTY);
+            }
+
+            return ACTIVE;
+        } catch (IOException e) {
+            return STALE;
+        }
+    }
+
+    @Override
+    protected void processRead(byte[] buffer) throws IOException {
+        if (reader == null) {
+            super.processRead(buffer);
+        } else {
+            final Message<?> message = reader.read();
+            if (message != null) {
+                handleMessage(this, message);
+            }
+        }
+    }
+
+    public void handshake(Request request) throws IOException {
+        final String version = request.getHeader(WebSocketHeaders.VERSION);
+        if (!"13".equals(version)) {
+            throw new VersionException(version);
+        }
+        if (request.getMethod() != Request.METHOD_GET) {
+            throw new HandshakeException("only GET method supported");
+        }
+        if (request.getHeader(WebSocketHeaders.KEY) == null) {
+            throw new HandshakeException("missing websocket key");
+        }
+        if (!WebSocketHeaders.isUpgradableRequest(request)) {
+            throw new HandshakeException("missing upgrade header");
+        }
+        sendHandshakeResponse(request);
+    }
+
+    public void sendMessage(Message<?> message) throws IOException {
+        if (writer == null) {
+            throw new IllegalStateException("websocket message was sent before handshake");
+        }
+        writer.write(message);
+    }
+
+    protected void handleMessage(WebSocketSession session, Message<?> message) throws IOException {
+        if (message instanceof PingMessage) {
+            server.handleMessage(session, (PingMessage) message);
+        } else if (message instanceof PongMessage) {
+            server.handleMessage(session, (PongMessage) message);
+        } else if (message instanceof TextMessage) {
+            server.handleMessage(session, (TextMessage) message);
+        } else if (message instanceof BinaryMessage) {
+            server.handleMessage(session, (BinaryMessage) message);
+        } else if (message instanceof CloseMessage) {
+            server.handleMessage(session, (CloseMessage) message);
+        }
+    }
+
+    @Override
+    public void handleException(Throwable e) {
+        log.error(e);
+        if (e instanceof WebSocketException) {
+            close(((WebSocketException) e).code());
+            return;
+        }
+        super.handleException(e);
+    }
+
+    public void close(short code) {
+        try {
+            sendMessage(new CloseMessage(code));
+        } catch (Exception e) {
+            log.warn("error while sending closing frame", e);
+        } finally {
+            close();
+        }
+    }
+
+    @Override
+    public void close() {
+        try {
+            extensions.forEach(Extension::close);
```

Comment:
Is it possible to get some exception from Extension.close and stay with some unclosed ones?

Reply:
fixed